### PR TITLE
Issue #8328: Change name of checkstyle types to be close to real java types - basic types

### LINKED ISSUE #8328: Change name of checkstyle types to match real java types
as we agreed at https://github.com/checkstyle/checkstyle/pull/8305#discussion_r436750608
discussed at https://github.com/checkstyle/checkstyle/issues/8311

we need to change names of our types in html(xdoc) documentations, it will not be change in functionality, just make naming for exact.

https://checkstyle.org/property_types.html

**Types:**
| old name  | new name  | Comment  |
|---        |---        |---        |
| integer   |  int,Integer,float,.....   | we create bunch of new types that we actually support, all primitives and their java types.    |
| string    | String    | in javadoc name is java.lang.String, in xdoc should be short `String`  |
| boolean   | boolean,Boolean   | in javadoc name should java.lang.Boolean, in xdoc should be `Boolean`  |
| stringSet  | String[]  | It is comma separated and trimmed Strings. [RelaxedStringArrayConverter](https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L358)  |
|  intSet | char[], byte[], short[], int[], float[],double[]  | we should be be clear what Check is actually using in its setter  |
| regexp   | Pattern  | there special [converter](https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L298), but does simple `java.util.regex.Pattern` instantiation. in javadoc `java.util.regex.Pattern` in xdoc `Pattern`  |
| uri   |  Uri | in description of type .... as there is special [converter](https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L331) with custom logic on priority to match resource, details are at [doc](https://checkstyle.org/property_types.html#uri)  |
|  file |  File |  let's keep it, we can not come to agreement on it, it is used only ones at Checker.cacheFile, and actual type of this property is `String`, [code](https://github.com/checkstyle/checkstyle/blob/c42a0e1b63c33400e05c98b62179cf303724e525/src/main/java/com/puppycrawl/tools/checkstyle/Checker.java#L148) |
| lineSeparator  |  LineSeparatorOption | Check do conversion to enum LineSeparatorOption by 'Enum.valueOf', javadoc should have fully qualified name, xdoc short name of enum  |
| parenPad   |  PadOption | com.puppycrawl.tools.checkstyle.checks.whitespace.PadOption name is javadoc and short name in xdoc  |
| wrapOp  | WrapOption  |  com.puppycrawl.tools.checkstyle.checks.whitespace.WrapOption name in javadoc and short name in Xdoc |
| block  | BlockOption  | com.puppycrawl.tools.checkstyle.checks.blocks.BlockOption name in javadoc and BlockOption in xdoc  |
| lcurly  | LeftCurlyOption  |  com.puppycrawl.tools.checkstyle.checks.blocks.LeftCurlyOption name in javadoc and BlockOption in xdoc |
| rcurly  | RightCurlyOption  | com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyOption name in javadoc and BlockOption in xdoc  |
| scope  |  Scope | in javadoc new name `com.puppycrawl.tools.checkstyle.api.Scope` in xdoc `Scope`. even there is custom converter, it does simple instantiation of Scope enum.  |
| access_modifiers  | AccessModifierOption[]  |  [special convert](https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L383) is used  |
| severity  | SeverityLevel  | in javadoc `com.puppycrawl.tools.checkstyle.api.SeverityLevel` in xdoc SeverityLevel. there is custom converter, it does simple instantiation of SeverityLevel enum. |
| importOrder  |  ImportOrderOption |  com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderOption name is javadoc and short name in xdoc |
| elementStyle | ElementStyleOption  |  com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationUseStyleCheck.ElementStyle name is javadoc and short name in xdoc |
| closingParens  | ClosingParensOption  |  com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationUseStyleCheck.ClosingParens name is javadoc and short name in xdoc |
| trailingArrayComma  | TrailingArrayCommaOption  | com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationUseStyleCheck.TrailingArrayComma name is javadoc and short name in xdoc  |
| javadocContentLocation  |  JavadocContentLocationOption | com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocContentLocationOption name is javadoc and short name in xdoc |

We should update https://checkstyle.org/property_types.html to mention exact name of classes and probably event provide link to source.

We should sort type alphabetically to ease search by users (as last PR)

For enums we will be honest  - type is `String. Value of ClosingParens`.
`tokens` - `String[]. Subset of tokens .....` [code](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L95).

In metadata we can put extra property `values-provider` and set it as enum type fully qualified name, to let plugins get all possible values. Property `tokens` should be special property and plugins should know about it and get all values from Check instance.

**Attention**: AnnotationUseStyleCheck has 3 inner enums that are listed here.

converters are at https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L117

custom types converters - https://github.com/checkstyle/checkstyle/blob/2e8c277c181487d81231170ff4b959edc008114e/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L163

### PR DESCRIPTION
**Pull request by gaurabdg**
#8328 

Refactored
1. `integer -> Integer`
2. `boolean -> Boolean`
3. `integerSet -> int[]`
4. create `Double` and `double[]`
5. `stringSet -> String[]`

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java [hunk 1]
```diff
@@ -59,20 +59,20 @@
...
- * {@code id} - a <a href="https://checkstyle.org/property_types.html#string">string</a>
+ * {@code id} - a <a href="https://checkstyle.org/property_types.html#String">String</a>
  * matched against the <a href="https://checkstyle.org/config.html#Id">check id</a>
  * associated with an audit event.
  * Optional as long as {@code checks} or {@code message} is specified.
  * </li>
  * <li>
  * {@code lines} - a comma-separated list of values, where each value is an
- * <a href="https://checkstyle.org/property_types.html#integer">integer</a>
+ * <a href="https://checkstyle.org/property_types.html#Integer">integer</a>
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java:69]: integer -> Integer

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java [hunk 2]
```diff
@@ -59,20 +59,20 @@
...
  * or a range of integers denoted by integer-integer.
  * It is optional.
  * </li>
  * <li>
  * {@code columns} - a comma-separated list of values, where each value is an
- * <a href="https://checkstyle.org/property_types.html#integer">integer</a>
+ * <a href="https://checkstyle.org/property_types.html#Integer">integer</a>
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java:75]: same here

File: src/xdocs/config_filters.xml [hunk 3]
```diff
@@ -513,21 +513,21 @@ public class UserService {
...
+            a <a href="property_types.html#String">String</a>
             matched against the <a href="config.html#Id">check id</a> associated with an audit
             event. Optional as long as <code>checks</code> or <code>message</code>
             is specified.
           </li>
           <li>
             <code>lines</code> - a comma-separated list of
             values, where each value is
-            an <a href="property_types.html#integer">integer</a> or a
+            an <a href="property_types.html#Integer">integer</a> or a
```
Comment (strkkk) [inline src/xdocs/config_filters.xml:524]: same here

File: src/xdocs/config_filters.xml [hunk 4]
```diff
@@ -513,21 +513,21 @@ public class UserService {
...
             range of integers denoted by integer-integer. It is optional.
           </li>
           <li>
             <code>columns</code> - a comma-separated list of
             values, where each value is
-            an <a href="property_types.html#integer">integer</a> or a
+            an <a href="property_types.html#Integer">integer</a> or a
```
Comment (strkkk) [inline src/xdocs/config_filters.xml:530]: same here

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java (hunk 1 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java (hunk 2 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionFilter.java]: Done

File: src/xdocs/config_filters.xml (hunk 3 above)
Comment (gaurabdg) [reply src/xdocs/config_filters.xml]: Done

File: src/xdocs/config_filters.xml (hunk 4 above)
Comment (gaurabdg) [reply src/xdocs/config_filters.xml]: Done

File: src/xdocs/property_types.xml [hunk 5]
```diff
@@ -30,21 +30,28 @@
...
       <p>
         This type represents an integer. The string representation is
         parsed using the <code>java.lang.Integer</code> class.
       </p>
     </section>
 
-    <section name="string">
+    <section name="Double">
+      <p>
+        This type represents an integer. The string representation is
```
Comment (rnveach) [inline src/xdocs/property_types.xml:42]: Double is not an integer. Integers have no decimal points, doubles do.

File: src/xdocs/property_types.xml [hunk 6]
```diff
@@ -30,21 +30,28 @@
...
       <p>
         This type represents a string. The literal string representation
         is used.
       </p>
     </section>
 
-    <section name="boolean">
+    <section name="Boolean">
       <p>
         This type represents a boolean. The default value is <code>false</code>.
```
Comment (rnveach) [inline src/xdocs/property_types.xml:56]: @romani Isn't `The default value` misleading? Default value is determined by check, and we don't always default to `false`. I thought this was more for the string representation, but we have a statement later covering that.

File: src/xdocs/property_types.xml [hunk 7]
```diff
@@ -100,6 +107,27 @@
       </source>
     </section>
 
+    <section name="double[]">
+      <p>
+        This type represents a set of integers. The string representation
```
Comment (rnveach) [inline src/xdocs/property_types.xml:112]: Same as before.

File: src/xdocs/property_types.xml (hunk 5 above)
Comment (gaurabdg) [reply src/xdocs/property_types.xml:42]: Sorry for that, forgot to change after copying the Integer description

File: src/xdocs/property_types.xml (hunk 6 above)
Comment (romani) [reply src/xdocs/property_types.xml:56]: yes, good catch.
@gaurabdg , please remove this line in separate commit or PR.

File: src/xdocs/config_design.xml [hunk 8]
```diff
@@ -878,14 +878,14 @@ public class Foo { // OK, only one top-level class
             <tr>
               <td>max</td>
               <td>Specify maximum allowed number of throws statements.</td>
-              <td><a href="property_types.html#integer">Integer</a></td>
+              <td><a href="property_types.html#Integer">Integer</a></td>
```
Comment (romani) [inline src/xdocs/config_design.xml:881]: please use exact type from setter.
it is `int`
https://github.com/checkstyle/checkstyle/blob/ae324ed2d8eebcdfc8f164095de11b7ae6714a32/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/ThrowsCountCheck.java#L230

please extend page that have list of all types to reference all types we support. We just should be exact on what module is actually has. We unhide extra detail only for enums to let people know possible values, but still name is exact name of enum type.

File: src/xdocs/property_types.xml (hunk 7 above)
Comment (gaurabdg) [reply src/xdocs/property_types.xml]: Done

File: src/xdocs/config_design.xml (hunk 8 above)
Comment (gaurabdg) [reply src/xdocs/config_design.xml:881]: Setting the exact type of setter will be highly inconsistent, especially for `Pattern`  and `Pattern[]`.
For example, `staticGroups` has filed class `Pattern[]`, but setter type `String...`
https://github.com/checkstyle/checkstyle/blob/ae324ed2d8eebcdfc8f164095de11b7ae6714a32/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L416
https://github.com/checkstyle/checkstyle/blob/ae324ed2d8eebcdfc8f164095de11b7ae6714a32/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L514

The field Class is more accurate description for a property I guess.

File: src/xdocs/property_types.xml (hunk 6 above)
Comment (gaurabdg) [reply src/xdocs/property_types.xml:56]: Done.

File: src/xdocs/config_design.xml (hunk 8 above)
Comment (romani) [reply src/xdocs/config_design.xml:881]: We should not go in internals of Checks. Inyernal field could change at any time and it should not change type that users set. 

This `String... packageGroups) ` should be `String[]` as it is array, that split by `,` and later each element converted to Pattern.  If user try to use `,` as part of regexp he will experience problems. 

So it is better to tell user a truth. Even for enums we do `String value of BlockOption` name to make it transperet it is a String and there is possible values in BlockOption

**Comment by romani** [comment]
Looks like update will be tricky  ... let remove from this PR update for `String[]` we will do this separately, with special attention to each property. Please try to do most simple update sooner, and we will left for later some controversial cases.

Please send PR with update of option names to make all of them end with "Option".

Please create one more PR to update https://checkstyle.org/property_types.html and reference all basic types (int, char, ..... , int[], char[], ....), we support them so no matter is Checks have them or not they should be referenced.

File: src/xdocs/releasenotes_old.xml [hunk 9]
```diff
@@ -274,8 +274,8 @@
         </li>
         <li>
           Enhanced the property types
-          <a href="property_types.html#stringSet">stringSet</a> and
-          <a href="property_types.html#intSet">intSet</a> to support being
+          <a href="property_types.html#String[]">String[]</a> and
```
Comment (romani) [inline src/xdocs/releasenotes_old.xml:277]: please encode `[]`  at `#String[]` links should work in browser.

Comment (gaurabdg) [reply src/xdocs/releasenotes_old.xml]: Sorry, my bad, done.

File: src/xdocs/config_design.xml (hunk 8 above)
Comment (romani) [reply src/xdocs/config_design.xml]: it is done.

**Pull request merged**